api/comms/chat.go

```diff
@@ -0,0 +1,454 @@
+	SET is_hidden = NOT EXISTS(
+
+		-- Check for chat messages
+		SELECT msg.message_id
+		FROM chat_message msg
+		LEFT JOIN chat_blast b USING (blast_id)
+		WHERE msg.chat_id = member.chat_id
+		AND (cleared_history_at IS NULL OR msg.created_at > cleared_history_at)
+		AND (msg.blast_id IS NULL OR b.from_user_id != member.user_id)
+
+		UNION
+
+		-- Check for chat message reactions
+		SELECT r.message_id
+		FROM chat_message_reactions r
+		LEFT JOIN chat_message msg ON r.message_id = msg.message_id
+		WHERE msg.chat_id = member.chat_id
+		AND r.user_id != member.user_id
+		AND (cleared_history_at IS NULL OR (r.created_at > cleared_history_at AND msg.created_at > cleared_history_at))
+	),
+	unread_count = (
+		select count(*)
+		from chat_message msg
+		where msg.created_at > COALESCE(member.last_active_at, '1970-01-01'::timestamp)
+		and msg.user_id != member.user_id
+		and msg.chat_id = member.chat_id
+	)
+	WHERE member.chat_id = $1
+	`, chatId)
+	return err
+}
+
+func chatSendMessage(tx pgx.Tx, ctx context.Context, userId int32, chatId string, messageId string, messageTimestamp time.Time, ciphertext string) error {
+	var err error
+
+	_, err = tx.Exec(ctx, "insert into chat_message (message_id, chat_id, user_id, created_at, ciphertext) values ($1, $2, $3, $4, $5)",
+		messageId, chatId, userId, messageTimestamp, ciphertext)
+	if err != nil {
+		return err
+	}
+
+	// update chat's info on last message
+	err = chatUpdateLatestFields(tx, ctx, chatId)
+	if err != nil {
+		return err
+	}
+
+	// sending a message implicitly marks activity for sender...
+	err = chatReadMessages(tx, ctx, userId, chatId, messageTimestamp)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+func chatReactMessage(tx pgx.Tx, ctx context.Context, userId int32, chatId string, messageId string, reaction *string, messageTimestamp time.Time) error {
+	var err error
+	if reaction != nil {
+		_, err = tx.Exec(ctx, `
+		insert into chat_message_reactions
+			(user_id, message_id, reaction, created_at, updated_at)
+		values
+			($1, $2, $3, $4, $4)
+		on conflict (user_id, message_id)
+		do update set reaction = $3, updated_at = $4 where chat_message_reactions.updated_at < $4`,
+			userId, messageId, *reaction, messageTimestamp)
+	} else {
+		_, err = tx.Exec(ctx, "delete from chat_message_reactions where user_id = $1 and message_id = $2 and updated_at < $3", userId, messageId, messageTimestamp)
+	}
+	if err != nil {
+		return err
+	}
+
+	// update chat's info on reaction
+	err = chatUpdateLatestFields(tx, ctx, chatId)
+	return err
+}
+
+func chatReadMessages(tx pgx.Tx, ctx context.Context, userId int32, chatId string, readTimestamp time.Time) error {
+	_, err := tx.Exec(ctx, "update chat_member set unread_count = 0, last_active_at = $1 where chat_id = $2 and user_id = $3",
+		readTimestamp, chatId, userId)
+	return err
+}
+
+var permissions = []ChatPermission{
+	ChatPermissionFollowees,
+	ChatPermissionFollowers,
+	ChatPermissionTippees,
+	ChatPermissionTippers,
```

Comment:
do we have permission for coins yet? maybe not

Reply:
not yet